[sw, runtime] Introduce "DIF" like PMP library

silvestrst opened the pull request:
This PR introduces the PMP access library which follows design principles of DIFs (low level HW manipulation API that higher level code builds upon). For more information please see the descriptions in this PR commits.

At the moment only TOR mode is minimally tested.

TODOs:

- Test NA4.

- Add more function descriptions in the source file.

- Hook up to verilator CI tests.

- Return from the fault handler.

alistair23 commented:
The config logic all looks correct.

I agree that you should have helper functions for NAPOT. It can be tricky to get right so it's best to just do it all in one place.

There is a risk of splitting config and address functions in that an address could be calculated assuming NAPOT but then configured using TOR (as an example). It might be easier to do both at once. That way you can also enforce that the address is written before the lock bit is set.

I think it's reasonable to hard code the granularity and WARL fields.

The large switch statements are a pain point with PMP, but there isn't always a way around it. PMP now supports 64 regions which makes the switch statements less palatable though.

I think checking should be done by the OS itself, there is no need to hand over data. The lock bit still allows an M-mode OS to read the config so it can query the status. An S-Mode OS should just be told a memory region to use/not-use. Plus then you don't have to worry about defining a new format for an OS to understand.

When OT has a set PMP config we can add querying to Tock to adapt around that.

silvestrst commented:
@alistair23 thank you for the review! Only came around today to look at the comments.

> There is a risk of splitting config and address functions in that an address could be calculated assuming NAPOT but then configured using TOR (as an example). It might be easier to do both at once. That way you can also enforce that the address is written before the lock bit is set.

Initially I have done the address setting in the configure function. I agree that would be the best way if address setting was uniform for all modes. The problem is that for TOR we then are required to configure two addresses, which for other cases will mean passing one dummy address.

I think maybe instead of passing `pmp_region_addr_matching_mode_t` in configuration structure, simply creating 4 separate configuration routines:

silvestrst commented:
> The large switch statements are a pain point with PMP, but there isn't always a way around it. PMP now supports 64 regions which makes the switch statements less palatable though.

This suggests that I am reading an out-of-date specification, I thought there were only 16 PMP regions?

silvestrst commented:
> I think checking should be done by the OS itself, there is no need to hand over data.

Am I understanding correctly that you are suggesting dropping `pmp_config_csrs_get`?

lenary commented:
> > The large switch statements are a pain point with PMP, but there isn't always a way around it. PMP now supports 64 regions which makes the switch statements less palatable though.
> 
> This suggests that I am reading an out-of-date specification, I thought there were only 16 PMP regions?

The draft spec has 64 regions. Ibex continues to support a maximum of 16 regions

imphil commented:
see also https://github.com/lowRISC/ibex/issues/908

silvestrst commented:
@lenary and @imphil thanks! I will have a look, I think though the way this library is implemented - it will be trivial to bump it to support 64 regions.

alistair23 commented:
Another idea would be to call some address config functions multiple times so specify all of the addresses, then call a set function that actually writes it to the hardware.

```C
// No register writes
pmp_set_mode(TOR);
pmp_set_access(READ, 0x10000, 0x100);
pmp_set_access(WRITE | EXE 0x20000, 0x100);

// Write configuration to hardware
pmp_set(LOCK);
```

Yeah, I don't think you need `pmp_config_csrs_get()`

silvestrst commented:
Just pushed some changes - WIP (incomplete changes - non-working state).

lenary reviewed (commented):
This is a good start. It would be good to understand what you intend for the `configure_tor` function.

Hopefully the x-macros I have defined will work for what we want, and should be relatively easy to solve.

One worry I have with the implementation here is some of the call graphs are reasonably "deep" which may prevent inlining. With the x-macros, it might be easier to have separate functions for reading/writing an individual cfg/address CSR, which don't need to go all through the same function which includes a parameter as to whether to read/write.

lenary commented:
> I think checking should be done by the OS itself, there is no need to hand over data. The lock bit still allows an M-mode OS to read the config so it can query the status. An S-Mode OS should just be told a memory region to use/not-use. Plus then you don't have to worry about defining a new format for an OS to understand.

@alistair23 I don't know what you mean by this - this is a simple, DIF-like library, which will be used before we've booted an OS (and from other C code in the system).

I agree that the per-type API is a better approach, as it lets us calculate the `pmpaddr` value at the same time as setting up the `pmpcfg`. I want to avoid your proposed stateful API, which doesn't allow flexibility in the callers. 

I would like a C function that can read the relevant CSRs for a given region to find out how it has been configured, and translate that back to similar structures that were used to configure it. As with the DIFs, the idea would be any C code in the system that wants to deal with PMP (reading or writing the configuration) should go through this library (this requirement is less onerous than the DIFs, in that I don't think Tock will be required to use this library).

alistair23 commented:
> > 
> 
> @alistair23 I don't know what you mean by this - this is a simple, DIF-like library, which will be used before we've booted an OS (and from other C code in the system).

I was responding to the following bullet point in the PR:

> I added pmp_config_csrs_get function to get all raw values from the pmpcfg and pmpaddr CSRs, with the anticipation that at some point there will be a need for handover of information to an operating system. Something I am not sure about, because I think this should be done by the OS itself - checking current configuration to correctly configure the regions it needs.

I agree that setting the address and config at the same time seems pretty useful.

Reading it back out makes sense. I just wonder how many users there are for reading the PMP configuration back out.

silvestrst commented on sw/device/lib/runtime/pmp.c:
```diff
@@ -0,0 +1,379 @@
+  return address == (address & PMP_ADDRESS_ALIGNMENT_INVERTED_MASK);
+}
+
+static pmp_region_configure_result_t pmp_cfg_csr_rw(
+    pmp_region_index_t region, pmp_csr_access_type_t access, uint32_t *value) {
+  switch (region) {
```
Looks good, thanks! I am doing some other refactoring at the moment, so you might not see this change in the very next push, but I will add it eventually.

lenary replied:
There used to be a function that worked out which part of the `pmpfcgN` register was the bit that mattered for the current region… but I guess that's gone now.

silvestrst commented on sw/device/lib/runtime/pmp.h:
```diff
@@ -0,0 +1,273 @@
+ * @return `pmp_region_configure_result_t`.
+ */
+PMP_WARN_UNUSED_RESULT
+pmp_region_configure_result_t pmp_region_configure_napot(
+  pmp_region_index_t region, const pmp_region_config_t *config,
+  pmp_region_address_t address);
```
NA4 is not available if G < 0, so I think a sensible thing to do here keep the public API for both, but implement NA4 in terms of NAPOT + additional check.

Nice suggestion, thanks.

silvestrst replied:
There is the comment that requests this functionality later on, marking this as resolved (duplicate).

silvestrst commented:
> Reading it back out makes sense. I just wonder how many users there are for reading the PMP configuration back out.

I agree, instead I will just add a routine to check whether a region is locked or not - which I think could be useful.

silvestrst commented:
> One worry I have with the implementation here is some of the call graphs are reasonably "deep" which may prevent inlining.

@lenary , I am not sure that this would be too much of a problem - I imagine this code will be used primarily for configuration of the registers, and not in performance critical code (like context switching).

silvestrst commented:
The last push was an intermediate push, I have not yet resolved @lenary comments!

silvestrst commented:
Needs some tidying up, the only mode that has been tested is TOR.

lenary commented on sw/device/lib/runtime/pmp.h:
```diff
@@ -0,0 +1,312 @@
+ * @return `pmp_region_configure_na4_result_t`.
+ */
+PMP_WARN_UNUSED_RESULT
+pmp_region_configure_na4_result_t pmp_region_configure_na4(
+    pmp_region_index_t region, const pmp_region_config_t *config,
+    pmp_region_address_t address);
```
I suggested this was combined with `pmp_region_configure_napot`, as you can choose the logic of how to encode the correct setting based on the `size` parameter. This means a simpler interface, where callers don't necessarily have to care what their size is, as long as it's larger than 4-bytes (rather than a specific call for 4-bytes, and other calls for 8-bytes). What was your opinion on this suggestion?

silvestrst replied:
I will give it a try next time I am working on the PMP.

lenary commented on sw/device/lib/runtime/pmp.c:
```diff
@@ -0,0 +1,539 @@
+  return kPmpRegionConfigureOk;
+}
+
+pmp_region_configure_result_t pmp_cfg_mode_lock_status_get(
+    pmp_region_index_t region, pmp_region_lock_t *lock) {
```
This does not match the name in the header - it should be `pmp_region_lock_status_get`, I think?

silvestrst replied:
Yup, thanks.

silvestrst replied:
Done.

lenary commented:
About WARL/Locked Errors, we had a similar [discussion](https://github.com/lowRISC/opentitan/pull/2042#pullrequestreview-419358896) about this over the padctrl DIF, and it makes sense to follow that here (with the guidance adjusted to follow the RISC-V spec, rather than the IP RTL, in this case):
- Check a region is locked before configuring it, if it's locked, return an error then. 
- If you know something is not compatible from the specification, then return an error (i.e., error conditions that the spec has required are noted).
- Otherwise, it's not up to you to return an error, and proceed as if you're writing something legal, and the write just worked (ie, don't read out again afterwards, and check the value you read was what you wrote, like you're currently doing for warl). This was chosen (for pinmux at least) because it's not clear what action a user should do if they get a Warl error.

mcy commented on sw/device/lib/runtime/pmp.h:
```diff
@@ -0,0 +1,312 @@
+ * "Volume II: RISC-V Privileged Architectures V20190608-Priv-MSU-Ratified",
+ * "3.6.1 Physical Memory Protection CSRs".
+ */
+#define PMP_REGIONS_NUM 16
```
Do any of these absolutely need to be macros? If not, they should be constants.

silvestrst replied:
I don't mind converting them to constants, but I don't quite understand why?

lenary replied:
This is a prime example of something that can be done in a follow-up PR.

mcy commented on sw/device/tests/pmp_sanitytest.c:
```diff
@@ -0,0 +1,85 @@
+/**
+ * The buffer that is used for load/store access violation test.
+ */
+__attribute__((aligned(4))) static volatile char pmp_load_store_test_data[2048];
```
Can you stick a `//` after the `__attribute__` to force clang-format to format this somewhat more sanely? (I think this should cause clang-format to wrap things a bit more nicely...)

mcy commented on sw/device/tests/pmp_sanitytest.c:
```diff
@@ -0,0 +1,85 @@
+__attribute__((aligned(4))) static volatile char pmp_load_store_test_data[2048];
+
+void handler_lsu_fault(void) {
+  const char exc_msg[] = "Load/Store fault, mtval shows the fault address\n";
+  LOG_INFO("%s", exc_msg);
```
Is there a reason this isn't inlined into the LOG call...?

(Also, LOG implies a newline at the end.)

mcy commented on sw/device/lib/runtime/pmp.h:
```diff
@@ -0,0 +1,312 @@
+ */
+PMP_WARN_UNUSED_RESULT
+pmp_region_configure_napot_result_t pmp_region_configure_napot(
+    pmp_region_index_t region, const pmp_region_config_t *config,
```
To be consistent with the DIF APIs, `config` should be taken by value, not by pointer.

silvestrst replied:
Will do.

silvestrst replied:
Done.

mcy commented on sw/device/tests/pmp_sanitytest.c:
```diff
@@ -0,0 +1,85 @@
+}
+
+static void pmp_load_violation_test(void) {
+  volatile char load = pmp_load_store_test_data[1023];
```
You can just write `(void)pmp_load_store_test_data[1023];` without assigning it anywhere. Because it's volatile, the mere manifestation of the lvalue *should* force the compiler to emit a read.

(Either way, the `volatile` on `load` is unnecessary.)

silvestrst replied:
Good point, done.

mcy commented on sw/device/lib/runtime/pmp.c:
```diff
@@ -0,0 +1,539 @@
+#define PMP_CFG_CSR_MODE_MASK 0x3
+
+// TODO - explanation to why macro magic is necessary (csr must be const expr).
+#define PMP_CSR_WRITE(csr, var) asm volatile("csrw " #csr ", %0;" : : "r"(var))
```
Do we want something like this in `stdasm.h`, where we can have a single place to explain this macro magic?

silvestrst replied:
Possibly, we could pull out these and put into the `stdasm`. Another option would be to have a separate library with wrapped asm for functions like:
`mepc_get`
`mepc_set`
`mtvec_set`
`mtvec_get`

Subject to a different PR (either of those options).

mcy replied:
Let's do a separate PR; personally, I favor the macro version since it shouldn't be too hard to mock.

mcy commented on sw/device/lib/runtime/pmp.c:
```diff
@@ -0,0 +1,539 @@
+// TODO - X macros as suggested by @lenary.
+
+// TODO - description.
+static pmp_region_configure_result_t pmp_cfg_csr_rw(
```
Since you've defined your own version of warn_unused_result, it might be prudent to add ti to these helper functions.

silvestrst replied:
Sounds good.

silvestrst replied:
Done on all static functions that return value.

mcy commented on sw/device/lib/runtime/pmp.c:
```diff
@@ -0,0 +1,539 @@
+// TODO - short description.
+static bool pmp_cfg_permissions_set(pmp_region_permissions_t perm,
+                                    uint32_t *bitfield) {
+  switch (perm) {
```
This makes me feel that `perm` should be a bitflag value, rather than an enum. It would definitely simply the code (and, for callers, allow the nicer `R | X`-type API).

silvestrst replied:
True, I guess I can just check for an illegal combination inside this function.

silvestrst replied:
Leaving this decision for the follow-up.

mcy commented on sw/device/lib/runtime/pmp.c:
```diff
@@ -0,0 +1,539 @@
+
+// TODO - short description.
+static void pmp_cfg_mode_lock_set(pmp_region_lock_t lock, uint32_t *bitfield) {
+  if (lock == kPmpRegionLockLocked) {
```
Consider inlining the condition into `bitfield_bit32`, so that the source code is branchless.

silvestrst replied:
used ternary operator instead

mcy commented on sw/device/lib/runtime/pmp.c:
```diff
@@ -0,0 +1,539 @@
+ * @param address Conventional system address.
+ * @return `true` on success, `false` on failure.
+ */
+static bool pmp_address_aligned(pmp_region_address_t address) {
```
Consider: `pmp_is_address_aligned`, to make it clear that this is a pure function that checks a property.

Though since you do other alignment checks in this file, maybe just having a single function that does alignment checks makes more sense...

silvestrst replied:
Done.

mcy commented on sw/device/lib/runtime/pmp.c:
```diff
@@ -0,0 +1,539 @@
+    return kPmpRegionConfigureNapotBadAddress;
+  }
+
+  // Check if the `size` if Power Of Two.
```
I think `if` was meant to be `is a`?

silvestrst replied:
Ooops, thanks.

silvestrst replied:
Done.

mcy commented on sw/device/lib/runtime/pmp.c:
```diff
@@ -0,0 +1,539 @@
+  // the NAPOT address.
+  uint32_t size_inverted = ~size;
+
+  *napot_address = address | size_inverted;
```
Shouldn't this be `& ~size`?

silvestrst replied:
Ooops, I think you are right. Thanks.

silvestrst replied:
Done.

mcy commented on sw/device/lib/runtime/pmp.c:
```diff
@@ -0,0 +1,539 @@
+  uint32_t size_inverted = ~size;
+
+  *napot_address = address | size_inverted;
+  *napot_address &= size_mask;
```
This line seems unnecessary, since we already know that `address` is `size`-aligned?

silvestrst replied:
This, sets the `size - 1` bits that otherwise are `0`.

silvestrst replied:
Now, this one is supposed to be `|`.

silvestrst replied:
Done.

mcy commented on sw/device/lib/runtime/pmp.c:
```diff
@@ -0,0 +1,539 @@
+// "Volume II: RISC-V Privileged Architectures V20190608-Priv-MSU-Ratified",
+// "3.6.1 Physical Memory Protection CSRs",
+// "Figure 3.28: PMP configuration register format".
+#define PMP_CFG_CSR_R 0
```
If these don't need to be defines, make them constants (with clearer names, like `Read` rather than `R`.)

silvestrst replied:
These correspond directly to PMP specification, so I would prefer to keep them as they are (regarding the s/R/READ/).

I don't see any value in making these defines into constants.

mcy replied:
> I don't see any value in making these defines into constants.

Macros-as-constants are not a pattern we should be using when we can avoid it. Macros pollute the token namespace and are untyped, leading to awkwardness around integer promotion.

Consider cleaning these up in a followup.

Also, there's no reason to stick to the spec's names when we can come up with something more readable, and write something like the following:
```c
/**
 * Corresponds to the `PMP_CFG_CSR_R` PMP constant.
 */
static const bitfield_bit32_index_t kPmpReadBit = 0;
```

mcy commented on sw/device/lib/runtime/pmp.c:
```diff
@@ -0,0 +1,539 @@
+#define PMP_CFG_CSR_MODE_MASK 0x3
+
+// TODO - explanation to why macro magic is necessary (csr must be const expr).
+#define PMP_CSR_WRITE(csr, var) asm volatile("csrw " #csr ", %0;" : : "r"(var))
```
Also, a separate thought: as written, this code cannot be unit-tested. We should think about maybe using @alphan and I's "Mask ROM component mocks" as a potential model for mocking CSR IO?

This is for a followup, obviously, but I wanted to start thinking about it.

silvestrst replied:
I think it can be unittested through `pmp_cfg_csr_rw` and `pmp_address_csr_rw`, bare in mind these functions are subject to change as per @lenary X-macros pattern suggestion.

mcy replied:
You can't mock partially mock a C file with our mocking system, unfortunately, so we probably can't do it that way.

alphan commented on sw/device/lib/runtime/pmp.c:
```diff
@@ -0,0 +1,539 @@
+}
+
+pmp_region_configure_result_t pmp_region_configure_tor(
+    pmp_region_index_t region, const pmp_region_config_t *config,
```
It may be helpful to call this sthg like `end_region`.

silvestrst replied:
Makes sense, thanks :)

silvestrst replied:
Done.

alphan reviewed (changes requested):
Thank you @silvestrst!

silvestrst commented on sw/device/lib/runtime/pmp.h:
```diff
@@ -0,0 +1,312 @@
+ * @return `pmp_region_configure_result_t`.
+ */
+PMP_WARN_UNUSED_RESULT
+pmp_region_configure_result_t pmp_region_configure_tor(
```
Not sure whether I should match napot signature here (instead of 2 addresses, have address and size).

lenary replied:
Two addresses is fine for the moment :)

silvestrst reviewed (commented):
Thanks guys, just had a quick run through your comments, will fix them up when next time working on this PR.

silvestrst commented on sw/device/lib/runtime/pmp.h:
```diff
@@ -0,0 +1,273 @@
+                                              pmp_region_config_t *config);
+
+// TODO - inline helpers to construct, NAPOT and TOR addresses! To what extent
+//        we need it? Without bitmanip it seems slightly inefficient.
```
Done.

silvestrst commented on sw/device/lib/runtime/pmp.h:
```diff
@@ -0,0 +1,312 @@
+#include <stdint.h>
+
+/**
+ * RVISC-V PMP address matching modes.
```
Done.

silvestrst commented on sw/device/lib/runtime/pmp.h:
```diff
@@ -0,0 +1,312 @@
+ * "3.6.1 Physical Memory Protection CSRs".
+ */
+#define PMP_REGIONS_NUM 16
+#define PMP_CFG_MULTIREG_COUNT 4
+#define PMP_ADDR_REG_COUNT 16
```
Done.

silvestrst commented on sw/device/lib/runtime/pmp.c:
```diff
@@ -0,0 +1,539 @@
+    return kPmpRegionConfigureNapotBadAddress;
+  }
+
+  // `size` is Power Of Two, so the invert can be used to clear the size bit of
+  // the NAPOT address.
+  uint32_t size_inverted = ~size;
+
+  *napot_address = address | size_inverted;
+  *napot_address &= size_mask;
```
Thank you for pointing it out, you are right, all addresses need to be right shifted 2 bits to the right and NAPOT 3 bits as the minimal NAPOT size is 8.

silvestrst replied:
This resulted in some minor refactoring to make it comprehensive, namely addition of `pmp_address_construct` and `PMP_ADDRESS_SHIFT_NAPOT`.

silvestrst replied:
Thank you for picking this up.

Done.

silvestrst commented on sw/device/lib/runtime/pmp.c:
```diff
@@ -0,0 +1,539 @@
+
+  // Check if the address is aligned to the `size`.
+  uint32_t size_mask_inverted = ~size_mask;
+  if (address == (address & size_mask_inverted)) {
```
done.

silvestrst commented:
**Still going through review comments, I expect to finish the easy ones next week, and do the more involved suggestions in the follow-up. When I am satisfied with the initial implementation, I will ask people to re-review this change quickly and create an issue with follow-up description**

- Addressed most of the easier comments
- Fixed several serious bugs
- Added NAPOT sanity test
- Tested TOR and NAPOT (both work - not extensive testing, see the sanity tests)
- Minor refactoring

silvestrst commented on sw/device/lib/runtime/pmp.h:
```diff
@@ -0,0 +1,312 @@
+ * Please see the PMP specification reference and general information at the
+ * top of this header file.
+ *
+ * Note: this function will clears PMP `region` configuration.
```
Done.

silvestrst commented on sw/device/lib/runtime/pmp.h:
```diff
@@ -0,0 +1,312 @@
+ *
+ * The values have been taken from the:
+ * "Volume II: RISC-V Privileged Architectures V20190608-Priv-MSU-Ratified",
+ * "3.6 Physical Memory Protection", section "Address Matching".
```
Indeed.

Done.

silvestrst replied:
Done.

silvestrst commented on sw/device/lib/runtime/pmp.h:
```diff
@@ -0,0 +1,312 @@
+ * effect in the "M" mode.
+ */
+typedef enum pmp_region_permissions {
+  kPmpRegionPermissionsNone = 0,         /**< No access permitted .*/
+  kPmpRegionPermissionsReadOnly,         /**< Only read access. */
+  kPmpRegionPermissionsExecuteOnly,      /**< Only execute access. */
+  kPmpRegionPermissionsReadExecute,      /**< Read and execute access. */
+  kPmpRegionPermissionsReadWrite,        /**< Read and write access. */
+  kPmpRegionPermissionsReadWriteExecute, /**< Read, write and execute access. */
```
Yes, that is correct. Miguel has also pointed it out that this could be done with bitfields. I can always check for an illegal combination in the source instead.

I think I will leave this refactor for the follow-up.

silvestrst commented on sw/device/lib/runtime/pmp.c:
```diff
@@ -0,0 +1,539 @@
+// "Volume II: RISC-V Privileged Architectures V20190608-Priv-MSU-Ratified",
+// "3.6.1 Physical Memory Protection CSRs",
+// "Figure 3.28: PMP configuration register format".
+#define PMP_CFG_CSR_R 0
+#define PMP_CFG_CSR_W 1
+#define PMP_CFG_CSR_X 2
+#define PMP_CFG_CSR_A 3
+#define PMP_CFG_CSR_L 7
```
I think that is a good idea, I should address it in the follow-up.

silvestrst commented on sw/device/tests/pmp_sanitytest.c:
```diff
@@ -0,0 +1,85 @@
+ * "3.6 Physical Memory Protection", "Address Matching":
+ *
+ * "If PMP entry i’s A field is set to TOR, the entry matches any address y
+ * such that pmpaddri−1 ≤ y < pmpaddri. If PMP entry 0’s A field is set to TOR,
```
I think that symbol is part of extended ASCII, but ill change it just to be on the safe side.

silvestrst replied:
Done.

silvestrst commented on sw/device/lib/runtime/pmp.c:
```diff
@@ -0,0 +1,539 @@
+// TODO - X macros as suggested by @lenary.
+
+// TODO - description.
+static pmp_region_configure_result_t pmp_cfg_csr_rw(
```
It would then require 2x functions with a switch statement. @lenary has a pretty good solution with X-macros, which I intend to fix-up in the fallow-up.

alphan replied:
Sounds good, we can continue this conversation once this starts using x-macros.

silvestrst commented on sw/device/lib/runtime/pmp.c:
```diff
@@ -0,0 +1,539 @@
+    return kPmpRegionConfigureError;
+  }
+
+  if (address_shifted != addr_csr_after_write) {
+    return kPmpRegionConfigureWarlError;
+  }
```
Good, point, @lenary has also raised that - and this has actually already been discussed in the DIF Padctrl.

silvestrst commented on sw/device/lib/runtime/pmp.c:
```diff
@@ -0,0 +1,539 @@
+// TODO - short description.
+static bool pmp_cfg_permissions_set(pmp_region_permissions_t perm,
+                                    uint32_t *bitfield) {
+  switch (perm) {
```
What do you mean? `pmp_cfg_permissions_set` and `pmp_cfg_mode_lock_set` are used to "accumulate" the value which eventually is passed to `pmp_csr_cfg_field_write`, the idea is that these functions help to build the final `pmpcfg` value.

alphan replied:
This is a minor point but it doesn't look like `pmp_cfg_permissions_set` is used the way you described. the value is always set to zero before it is called and the members of the enums are not individual bits in a bitfield, hence my comment.

silvestrst commented on sw/device/lib/runtime/pmp.c:
```diff
@@ -0,0 +1,539 @@
+    return kPmpRegionConfigureNapotBadAddress;
+  }
+
+  // Check if the `size` if Power Of Two.
```
Done.

silvestrst commented:
# I propose to merge this PR with all the outstanding comments and other actions resolved in the follow-up. Please see #3513 .

@mcy @alphan @lenary please "approve" this change if you are happy with the proposed strategy.

mcy commented on sw/device/lib/runtime/pmp.c:
```diff
@@ -0,0 +1,526 @@
+  return true;
+}
+
+// TODO - description.
```
I assume you're not merging with these TODOs? Can you please resolve these before merging?

silvestrst replied:
Done.

mcy commented on sw/device/tests/pmp_sanitytest_napot.c:
```diff
@@ -0,0 +1,110 @@
+#include "sw/device/lib/testing/test_main.h"
+#include "sw/device/lib/testing/test_status.h"
+
+#define PMP_LOAD_REGION_ID 0
```
Some of these also don't need to be macros. Please convert them to constants in a followup.

silvestrst replied:
I will do this in a follow-up.

mcy commented on sw/device/tests/pmp_sanitytest_napot.c:
```diff
@@ -0,0 +1,110 @@
+bool test_main(void) {
+  pmp_load_exception = false;
+  char load = pmp_load_store_test_data[PMP_LOAD_RANGE_BOTTOM_OFFSET];
+  CHECK(pmp_load_exception == false,
```
Don't write `cond == false`, write `!cond`.

silvestrst replied:
Done.

mcy commented on sw/device/tests/pmp_sanitytest_napot.c:
```diff
@@ -0,0 +1,110 @@
+
+  pmp_load_exception = false;
+  load = pmp_load_store_test_data[PMP_LOAD_RANGE_BOTTOM_OFFSET];
+  CHECK(pmp_load_exception == true,
```
`== true` is redundant.

silvestrst replied:
Done.

mcy commented on sw/device/tests/pmp_sanitytest_napot.c:
```diff
@@ -0,0 +1,110 @@
+  uintptr_t load_range_start =
+      (uintptr_t)&pmp_load_store_test_data[PMP_LOAD_RANGE_BOTTOM_OFFSET];
+
+  pmp_region_config_t config = {
+      .lock = kPmpRegionLockLocked, .permissions = kPmpRegionPermissionsNone,
+  };
```
Consider inlining this.

silvestrst replied:
I think I prefer not inlining it in this case.

mcy commented on sw/device/tests/pmp_sanitytest_napot.c:
```diff
@@ -0,0 +1,110 @@
+/**
+ * The buffer that is used for load/store access violation test.
+ */
+__attribute__((aligned(PMP_LOAD_RANGE_SIZE))) static volatile char
```
Consider following the attribute with // to force clang-format to wrap this.

silvestrst replied:
Done.

mcy commented on sw/device/lib/runtime/pmp.h:
```diff
@@ -0,0 +1,306 @@
+ * effect in the "M" mode.
+ */
+typedef enum pmp_region_permissions {
+  kPmpRegionPermissionsNone = 0,         /**< No access permitted .*/
```
You can clean this up in a followup, but please don't add new `/**<` comments, for consistency with existing code.

silvestrst replied:
I will do this in a follow-up.

mcy reviewed (approved):
Quite a few of my comments (mostly around todos and minor formatting) should be resolved before merging.

silvestrst commented:
@lenary could you please have a look? Merging this PR at the end of today otherwise.

lenary reviewed (approved):
One comment about mocking, but this looks good and the sanity tests are good too!
